### PR TITLE
More GPIO control through CLI

### PR DESCRIPTION
**Pull request by SamuelYvon**
# What's new

- Add a `gpio` command in the CLI to perform various GPIO operations (not limited to write)
- This is related to feature request https://github.com/flipperdevices/flipperzero-firmware/issues/1240

There are a few changes I'm expecting to maybe make before this is merged in:
- I used a goto to cleanup the string_t at the end of the function. However, I have noticed that nowhere else in the furi codebase this pattern is used (except the stmlibs). Is this against your coding guidelines?
- I left the previously available `gpio_set` for "retrocompatibility" but I am not sure this matters here? I could remove it.
- I would have liked to reset all the pins to input mode when the CLI session is finished. Is there a way to hook up to some event? I was not able to find anything.

# Verification 

- Using the interface for GPIO read / write / mode set should be enough

# Checklist (For Reviewer)

- [ ] PR has description of feature/bug or link to Confluence/Jira task
- [ ] Description contains actions to verify feature/bugfix
- [ ] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: applications/cli/cli_commands.c [hunk 1]
```diff
@@ -248,43 +248,195 @@ void cli_command_led(Cli* cli, string_t args, void* context) {
...
+
+        if(LL_GPIO_IsInputPinSet(gpio[num].port, gpio[num].pin)) {
+            printf("1");
+        } else {
+            printf("0");
+        }
+    }
+
+// TODO: no other goto in the code base
+// except in /lib. Is this not allowed?
```
Comment (skotopes) [inline applications/cli/cli_commands.c]: Not allowed.

File: applications/cli/cli_commands.c [hunk 2]
```diff
@@ -248,43 +248,195 @@ void cli_command_led(Cli* cli, string_t args, void* context) {
...
+        } else if(pin_action == set) {
+            printf("Wrong 3rd argument. Use \"0\" to set to low. Use \"1\" to set to high.");
+        } else {
+            printf("Wrong 3rd argument. Use \"0\" to set pin as output. Use \"1\" to set as input.");
+        }
+    }
+
+    if(pin_action == mode) {
+        if(value == 0) { // output
+            LL_GPIO_SetPinMode(gpio[num].port, gpio[num].pin, LL_GPIO_MODE_OUTPUT);
```
Comment (skotopes) [inline applications/cli/cli_commands.c]: No need in LL here, use furi_hal_gpio + furi_hal_resources

File: applications/cli/cli_commands.c [hunk 3]
```diff
@@ -248,43 +248,195 @@ void cli_command_led(Cli* cli, string_t args, void* context) {
...
+static void print_valid_pins(void) {
+    printf("Wrong pin name. Available pins: ");
+    for(uint8_t i = 0; i < sizeof(pin_names) / sizeof(char*); i++) {
+        printf("%s ", pin_names[i]);
+    }
+}
+
+void cli_command_gpio_ctl(Cli* cli, string_t args, void* context) {
+    UNUSED(context);
+    typedef enum GpioCtlAction_Enum {
```
Comment (skotopes) [inline applications/cli/cli_commands.c]: That's not how we do enums

File: applications/cli/cli_commands.c [hunk 4]
```diff
@@ -248,43 +248,195 @@ void cli_command_led(Cli* cli, string_t args, void* context) {
...
+    GpioCtlAction pin_action;
+    if(!string_cmp(action, "mode")) {
+        pin_action = mode;
+    } else if(!string_cmp(action, "set")) {
+        pin_action = set;
+    } else if(!string_cmp(action, "read")) {
+        pin_action = read;
+    } else {
+        printf("Wrong 2nd argument (%s). Use \"mode\" to change the pin mode, \"set\" to set the value of an output pin or \"read\" to read the value of an input pin.", string_get_cstr(action));
+        goto cli_command_gpio_ctl_cleanup;
```
Comment (skotopes) [inline applications/cli/cli_commands.c]: We don't do goto

**Review by skotopes — CHANGES_REQUESTED** [review]

File: applications/cli/cli_commands.c (hunk 1 above)
Comment (SamuelYvon) [reply applications/cli/cli_commands.c]: OK, I'll reimplement w/o gotos.

**Comment by SamuelYvon** [comment]
@skotopes , I'll reimplement w/o gotos. Regarding the existing `gpio-set` command, would you rather I remove it, or implement all three options as different commands (ie, gpio-set, gpio-mode, gpio-read). This would simplify the parsing for each command as well. I also noted your comment for `LL_` for the GPIO access. This is also how `gpio-set` is currently implemented. I'll refactor as part of the next iteration of this PR.

**Comment by SamuelYvon** [comment]
Please review the latest changes.

These changes follow from the review. Instead of having to
rely in GOTO to ensure cleaning, I divided in smaller pieces. Since I had no feedback w.r.t
the previously existing `gpio_set`, this fits more with the
current implementation.

I also tried to interpret "That's not how we do enums" as
to not include a `enum EnumName` decl. when using typedefs.

I also distanced the implementation from using LL_GPIO
in favour of functions provided in `<furi_hal_gpio.h>`
but there was nothing to get the mode, so I stuck with
LL_GPIO for mode check.

**Review by skotopes — APPROVED** [review]

**Comment by skotopes** [comment]
Hi @SamuelYvon, sorry for delay. I'll approve this PR, but will rewrite it's code a little bit to match our standard cli command routine

**Pull request merged**